From b635b5c4233105f90fc3b47d00aeab99dbc03e12 Mon Sep 17 00:00:00 2001
From: xdBronch <51252236+xdBronch@users.noreply.github.com>
Date: Sat, 15 Nov 2025 14:53:32 -0500
Subject: [PATCH 064/133] sema: improve codegen of for loop safety checks
X-Developer-Signature: v=1; a=openpgp-sha256; l=1802; i=xtex@astrafall.org;
 h=from:subject; bh=KreILVHxeyty4Cq1WgiOo/oX37h71mw0Ges86urxz0I=;
 b=owGbwMvMwCW2U4Ij7wZL9ETG02pJDJn78pjPv9op7hRxyc1h0lO5pzylMcoFDJ5b3kRqh2S8v
 ZR9fmZJx0QWBjEuBksxRZYiwwZvVp10ftFl5bIwc1iZQIZIizQwMDAwsDDw5SbmlRrpGOmZahvq
 GRrqGOkYMXBxCsBU+1xlZFj7OyBFK641N/TGnZL//TvTvPQYfdTnf43/fC37anIksykjw427U6s
 KfjPvnymQLMZ2ov1p98TnM5ZqNxVZHbm9c/bBjbwA
X-Developer-Key: i=xtex@astrafall.org; a=openpgp;
 fpr=7231804B052C670F15A6771DB918086ED8045B91

(cherry picked from commit 29a1d0c06ff80836d744d2ab5db3626bf17fa0d1)
---
 src/Sema.zig | 13 +++++++++++--
 1 file changed, 11 insertions(+), 2 deletions(-)

diff --git a/src/Sema.zig b/src/Sema.zig
index 43f08562aa91..2dcf6be98fc7 100644
--- a/src/Sema.zig
+++ b/src/Sema.zig
@@ -4550,6 +4550,9 @@ fn zirForLen(sema: *Sema, block: *Block, inst: Zir.Inst.Index) CompileError!Air.
             // This argument is a range.
             const range_start = try sema.resolveInst(zir_arg_pair[0]);
             const range_end = try sema.resolveInst(zir_arg_pair[1]);
+            if (try sema.resolveDefinedValue(block, arg_src, range_start)) |start| {
+                if (try sema.valuesEqual(start, .zero_usize, .usize)) break :l range_end;
+            }
             break :l try sema.analyzeArithmetic(block, .sub, range_end, range_start, arg_src, arg_src, arg_src, true);
         };
         const arg_len = try sema.coerce(block, .usize, arg_len_uncoerced, arg_src);
@@ -4613,12 +4616,18 @@ fn zirForLen(sema: *Sema, block: *Block, inst: Zir.Inst.Index) CompileError!Air.
 
     // Now for the runtime checks.
     if (any_runtime and block.wantSafety()) {
+        var ok: Air.Inst.Ref = .none;
         for (runtime_arg_lens, 0..) |arg_len, i| {
             if (arg_len == .none) continue;
             if (i == len_idx) continue;
-            const ok = try block.addBinOp(.cmp_eq, len, arg_len);
+            const eq = try block.addBinOp(.cmp_eq, len, arg_len);
+            ok = if (ok != .none)
+                try block.addBinOp(.bool_and, ok, eq)
+            else
+                eq;
+        }
+        if (ok != .none)
             try sema.addSafetyCheck(block, src, ok, .for_len_mismatch);
-        }
     }
 
     return len;
-- 
2.52.0

